src/cpp/client/channel_cc.cc

```diff
@@ -67,46 +69,64 @@ class TagSaver final : public CompletionQueueTag {
 // Constantly watches channel connectivity status to reconnect a transiently
 // disconnected channel. This is a temporary work-around before we have retry
 // support.
-class ChannelConnectivityWatcher {
+class ChannelConnectivityWatcher : private GrpcLibraryCodegen {
  public:
   static void StartWatching(grpc_channel* channel) {
-    char* env = gpr_getenv("GRPC_DISABLE_CHANNEL_CONNECTIVITY_WATCHER");
-    bool disabled = gpr_is_true(env);
-    gpr_free(env);
-    if (!disabled) {
-      gpr_once_init(&g_connectivity_watcher_once_, InitConnectivityWatcherOnce);
-      gpr_mu_lock(&g_watcher_mu_);
+    if (!IsDisabled()) {
+      std::unique_lock<std::mutex> lock(g_watcher_mu_);
       if (g_watcher_ == nullptr) {
         g_watcher_ = new ChannelConnectivityWatcher();
       }
       g_watcher_->StartWatchingLocked(channel);
-      gpr_mu_unlock(&g_watcher_mu_);
+    }
+  }
+
+  static void StopWatching() {
+    if (!IsDisabled()) {
+      std::unique_lock<std::mutex> lock(g_watcher_mu_);
+      if (g_watcher_->StopWatchingLocked()) {
+        delete g_watcher_;
+        g_watcher_ = nullptr;
+      }
     }
   }
 
  private:
-  ChannelConnectivityWatcher() {
+  ChannelConnectivityWatcher() : channel_count_(0), shutdown_(false) {
     gpr_ref_init(&ref_, 0);
     gpr_thd_options options = gpr_thd_options_default();
-    gpr_thd_options_set_detached(&options);
+    gpr_thd_options_set_joinable(&options);
     gpr_thd_new(&thd_id_, &WatchStateChange, this, &options);
   }
 
+  static bool IsDisabled() {
+    char* env = gpr_getenv("GRPC_DISABLE_CHANNEL_CONNECTIVITY_WATCHER");
+    bool disabled = gpr_is_true(env);
+    gpr_free(env);
+    return disabled;
+  }
+
   void WatchStateChangeImpl() {
     bool ok = false;
     void* tag = NULL;
     CompletionQueue::NextStatus status = CompletionQueue::GOT_EVENT;
     while (true) {
-      status = cq_.AsyncNext(&tag, &ok, gpr_inf_past(GPR_CLOCK_REALTIME));
-      // Make sure we've seen 2 TIMEOUTs before going to sleep
-      if (status == CompletionQueue::TIMEOUT) {
-        status = cq_.AsyncNext(&tag, &ok, gpr_inf_past(GPR_CLOCK_REALTIME));
-        if (status == CompletionQueue::TIMEOUT) {
-          gpr_sleep_until(
-              gpr_time_add(gpr_now(GPR_CLOCK_REALTIME),
-                           gpr_time_from_millis(kConnectivityCheckIntervalMsec,
-                                                GPR_TIMESPAN)));
-          continue;
+      {
```

Comment:
This extra scope is probably not needed, since it encompasses the whole body of the while loop.

Reply:
Thanks a lot for the review! The extra scope here is to release `shutdown_mu_` after the `if (shutdown_)` statement. It's not necessary to hold the lock in the whole body of the while loop.

Reply:
Ah!  Sorry, when I last looked at this, I thought that this scope was spanning the full scope of the loop.